### PR TITLE
Milestone#5: Refactor

### PR DESCRIPTION
**Pull request by Anjacodes**
In this milestone, we:

* Made sure YAGNI, KISS, and DRY rules and Ruby style guide are respected.
* Followed SOLID principles.
* Pay special attention to the "Single-responsibility Principle".
    * Made sure that the role of `main.rb` is only to start your code and ask the user for the initial choice of options (or to quit).
    * Extracted some pieces of logic into separate classes.

### CODE REVIEW

File: create_book.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+  def initialize(state)
+    @state = state
+    @book_list = @state[:book_list]
+  end
+
+  def create_book
+    print 'Title: '
+    title = gets.chomp
+    print 'Author: '
+    author = gets.chomp
```
Comment (TedLivist) [inline create_book.rb]: - [ ] In keeping with the single-responsibility rule of SOLID, the sole responsibility of the **create_book** method should be to simply create books. The user inputs and inputs validations should be handled by another method. Kindly extract these input requests into another method.

  **SUGGESTION:** You can extract these requests methods into another method and have that method return and array of title and author. Then destructure the variables in create_book like so:

  anja-book

File: create_person.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    @state = state
+  end
+
+  def create_person
+    print 'Do you want to create a student (1)  or a teacher (2)? [Input the number]: '
+    user_choice = gets.chomp
+    print 'Age: '
+    @age = gets.chomp.to_i
+    print 'Name: '
+    @name = gets.chomp
```
Comment (TedLivist) [inline create_person.rb]: - [ ] Kindly ensure that requests for user inputs are handled by another method in keeping with the single-responsibility rule.

File: create_person.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    when 1 then create_student
+    when 2 then create_teacher
+    else p 'Please enter either 1 or 2'
+    end
+  end
+
+  def create_student
+    print 'Has parent permission? [Y/N]: '
+    permission_input = gets.chomp.downcase
+    permission = permission_input == 'y'
```
Comment (TedLivist) [inline create_person.rb]: - [ ] Kindly extract this request into another method solely responsible for inputs and input validation

File: create_person.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    @state[:people_list] << Student.new(@age, @name, parent_permission: permission)
+    p 'Student created successfully'
+    puts ''
+  end
+
+  def create_teacher
+    print 'Specialization: '
+    specialization = gets.chomp
```
Comment (TedLivist) [inline create_person.rb]: - [ ] Kindly refactor into another method.

File: create_rental.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+      p "#{index}) Title: '#{book.title}', Author: #{book.author}"
+    end
+    book = @book_list[gets.chomp.to_i]
+    p 'Select a person from the following list by number (not id)'
+    @people_list.each_with_index do |person, index|
+      p "#{index}) [#{person.class.name}] Name: '#{person.name}', ID: #{person.id}, Age: #{person.age}"
+    end
+    person = @people_list[gets.chomp.to_i]
+    print 'Date: '
+    date = gets.chomp
```
Comment (TedLivist) [inline create_rental.rb:21]: - [ ] In keeping with the single-responsibility rule, kindly ensure these requests are handled by another method/class.

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @Anjacodes and @disc3110,

Good job so far!
✔️ Good documentation
✔️ Nice refactoring of code
However, there are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
- [ ] **In README file, I would suggest you define the actually that needs to be run (like `main.rb`)**

  anja-readme

- [ ] _Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: create_rental.rb (hunk 5 above)
Comment (Anjacodes) [reply create_rental.rb:21]: Hi @TedLivist, Thank you for the helpful feedback 🙏 . We have implemented and pushed all the suggested changes.

**Review by Mihndim2020 — APPROVED** [review]
# STATUS: APPROVED!!!  🎉 🎉 🎉 :champagne: :champagne: :champagne:
Hi @Anjacodes, 

Your project is complete! Great work on making the changes requested by the previous reviewer. There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 🎉 🎉 :champagne: :champagne: :champagne:

Congratulations-Memes-FEATURED

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**